docs/rfc/rfc-10.adoc

```diff
@@ -0,0 +1,248 @@
+"centrally" on Ethereum L1.
+
+There are two main design challenges. The first challenge is how to represent
+32-byte Solana addresses in the process of depositing and redeeming. The entire
+design of the Bridge is oriented around 20-byte Ethereum addresses. Bridge
+expects a 20-byte depositor address in the reveal function and passes 20-byte
+depositor addresses to the Bank when providing the sweep proof. The second
+challenge is optimistic minting. Only `TBTCVault` can mint tBTC and the
+optimistic minting has to happen in `TBTCVault`. Having another vault be
+responsible for Solana minting and bridging would mean we have to tightly couple
+vaults together, complicate optimistic minting code, or not allow for optimistic
+minting on Solana at all.
+
+The proposed design is oriented around a depositor contract that will reveal
+deposits to the bridge, map Solana addresses with revealed deposits, receive
+tBTC (optimistically minted or not), and perform bridging to Solana.
+
+==== P2SH deposit script
+
+The P2SH deposit script has to be enhanced to include optional extra data. The
+extra data will be used by depositors who are smart contracts. In the context of
+this proposal, we will use the extra data to associate the Bitcoin deposit with
+an address on a foreign (non-Ethereum) chain.
+
+```
+<depositor-address> DROP
+<depositor-extra-data> DROP
+<blinding-factor> DROP
+DUP HASH160 <signingGroupPubkeyHash> EQUAL
+IF
+  CHECKSIG
+ELSE
+  DUP HASH160 <refundPubkeyHash> EQUALVERIFY
+  <locktime> CHECKLOCKTIMEVERIFY DROP
+  CHECKSIG
+ENDIF
+```
+
+==== Reveal with depositor extra data
+
+`DepositRevealInfo` struct will gain an additional optional field. Since this
+structure is not persisted in the storage, the new field is not a concern for
+the contract upgrade:
+
+```
+/// @notice Represents data which must be revealed by the depositor during
+///         deposit reveal.
+struct DepositRevealInfo {
+    // (...)
+
+    // Optional extra data used by depositors who are smart contracts. The smart
+    // contract depositor can use the extra data to represent additional
+    // information, such as the depositor's address on a foreign (non-Ethereum)
+    // chain.
+    bytes32 depositorExtraData;        
+    // This struct doesn't contain `__gap` property as the structure is not
+    // stored, it is used as a function's calldata argument.
+}
+```
+
+The `DepositRequest` function will gain the same field. Since mapping values are
+stored in different slots we only need to make sure the new field comes after
+the existing ones to address upgradeability concerns.
+
+```
+/// @notice Represents tBTC deposit request data.
+struct DepositRequest {
+    // (...)
+
+    // Optional extra data used by depositors who are smart contracts. The smart
+    // contract depositor can use the extra data to represent additional
+    // information, such as the depositor's address on a foreign (non-Ethereum)
+    // chain.
+    bytes32 depositorExtraData;
```

Comment:
Will `bytes32` be enough? To be more generic don't we need to include a target chain ID and allow providing a public key (could be compressed format 33-bytes)?